pkg/kubelet/kubelet.go

```diff
@@ -2105,21 +2105,30 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
 	// TODO: out of resource eviction should have a pod admitter call-out
 	attrs := &lifecycle.PodAdmitAttributes{Pod: pod, OtherPods: pods}
 	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		// If pod resource allocation checkpoint manager (checkpointState) is nil, it is likely because
+		// kubelet is bootstrapping kube-system pods on master node. In this scenario, pod resource resize
+		// is neither expected nor can be handled. Fall back to regular 'canAdmitPod' processing.
+		// TODO(vinaykul,InPlacePodVerticalScaling): Investigate if we can toss out all this checkpointing
+		// code and instead rely on ResourceAllocation / Resize values persisted in PodStatus. (Ref: KEP 2527)
 		// Use allocated resources values from checkpoint store (source of truth) to determine fit
-		otherPods := make([]*v1.Pod, 0, len(pods))
 		checkpointState := kl.statusManager.State()
-		for _, p := range pods {
-			op := p.DeepCopy()
-			for _, c := range op.Spec.Containers {
-				resourcesAllocated, found := checkpointState.GetContainerResourceAllocation(string(p.UID), c.Name)
-				if c.Resources.Requests != nil && found {
-					c.Resources.Requests[v1.ResourceCPU] = resourcesAllocated[v1.ResourceCPU]
-					c.Resources.Requests[v1.ResourceMemory] = resourcesAllocated[v1.ResourceMemory]
+		if checkpointState != nil {
+			otherPods := make([]*v1.Pod, 0, len(pods))
+			for _, p := range pods {
+				op := p.DeepCopy()
+				for _, c := range op.Spec.Containers {
+					resourcesAllocated, found := checkpointState.GetContainerResourceAllocation(string(p.UID), c.Name)
+					if c.Resources.Requests != nil && found {
+						c.Resources.Requests[v1.ResourceCPU] = resourcesAllocated[v1.ResourceCPU]
+						c.Resources.Requests[v1.ResourceMemory] = resourcesAllocated[v1.ResourceMemory]
+					}
 				}
+				otherPods = append(otherPods, op)
 			}
-			otherPods = append(otherPods, op)
+			attrs.OtherPods = otherPods
+		} else {
+			klog.ErrorS(nil, "pod resource allocation checkpoint manager is not initialized.")
```

Comment:
If checkpoint state is nil in non-error cases (like standalone kubelet) we can't log errors on every pod sync. We'll flood the error log with these otherwise. 

Is there a way we can make VPA inactive/inert for standalone kubelets, since they won't be getting pod updates from the API that change resources?

Reply:
Thanks for the review and catching this issue. I think I can add a couple of Get interfaces to status manager (similar pattern to GetPodStatus) and avoid these nil checks. I'll work on it.

Reply:
BTW, how did you find this issue? Is this it: https://github.com/kelseyhightower/standalone-kubelet-tutorial ?

Reply:
> how did you find this issue?

we (GKE) run standalone kubelets, and our CI started failing once the VPA change merged to master

Reply:
> I think I can add a couple of Get interfaces to status manager (similar pattern to GetPodStatus) and avoid these nil checks

Does the VPA feature make sense for a standalone kubelet? I don't think it does. It would be easier to reason about if standalone kubelets didn't ever hit VPA code paths.

Reply:
Is there a way to repro locally on my mac (ubuntu VM)? (With my company shutting down, I don't have access to GKE anymore)

For the in-place pod resize part of the code, I can optimize away [invocation of handlePodResourcesResize](https://github.com/kubernetes/kubernetes/blob/0270fc75d0f59e1f5a31a810737ac35ac3eeb8a3/pkg/kubelet/kubelet.go#L1841L1843) based on an additional KubeClient==nil check. But, imho, the code that tracks ResourcesAllocated (upon pod addition) and Resize (which will always be "" in this case) should continue to work in its current form consistency with how GetPodStatus/SetPodStatus currently works. (Just by looking at code, they seem to be in play)

Reply:
maybe something like this:

```
mkdir -p /var/run/kubernetes/static-pods

echo "
kind: Pod
apiVersion: v1
metadata:
  name: busybox
spec:
  containers:
  - name: busybox
    image: busybox
" > /var/run/kubernetes/static-pods/busybox.yaml

START_MODE=kubeletonly FEATURE_GATES=InPlacePodVerticalScaling=true hack/local-up-cluster.sh
```

but I bet local-up-cluster.sh has atrophied and START_MODE=kubeletonly won't work quite right

Reply:
@SergeyKanzhelev @smarterclayton looks like we have a general test gap for standalone kubelet that regularly causes problems:
* https://github.com/kubernetes/kubernetes/pull/113501
* https://github.com/kubernetes/kubernetes/pull/108442
* https://github.com/kubernetes/kubernetes/issues/108063
* https://github.com/kubernetes/kubernetes/issues/87558
* https://github.com/kubernetes/kubernetes/issues/77174
* https://github.com/kubernetes/kubernetes/pull/75308 (this actually seems similar to this PR in keeping code paths active and trying to add nil checks, which exposed later code to unexpected nils)

Reply:
> 
> START_MODE=kubeletonly FEATURE_GATES=InPlacePodVerticalScaling=true hack/local-up-cluster.sh
> ```
> 
> but I bet local-up-cluster.sh has atrophied and START_MODE=kubeletonly won't work quite right

Thanks, that works. Needs one additional envvar: 
`POD_MANIFEST_PATH=/var/run/kubernetes/static-pods START_MODE=kubeletonly FEATURE_GATES=InPlacePodVerticalScaling=true hack/local-up-cluster.sh`

Anyways, I have another commitment I need to get to. I'll verify this with the above suggested optimization later in the day today.

Reply:
Thanks Jordan for raising this, AFAIU I don't think we have good coverage of standalone kubelet in CI today. I think we could extend node e2e to also support a mode where we spin up kubelet in standalone mode specifically and only run static pods as a few tests. 

In general, it may help as well to have an environment we can dedicate to static pod related tests since we have a few issues there that @smarterclayton and others been trying to tackle recently.

Reply:
Looking more closely, `kubeletonly` mode conveniently brings up just kubelet (POD_MANIFEST_PATH default works) but kubeClient is created. But, I was able to set `kubeDeps.KubeClient = nil` just before NewMainKubelet and get a repro to verify the before & after.

We don't need a new check for `kubeClient==nil` before invoking  handlePodResourcesResize  because `!kubetypes.IsStaticPod(pod)` check achieves the same goal.